### PR TITLE
Initial Topology Hints e2e Tests

### PR DESCRIPTION
**Pull request by robscott**
#### What type of PR is this?
None of these really make sense, but going with default:

/kind feature

#### What this PR does / why we need it:
This adds e2e tests for Topology Hints

#### Special notes for your reviewer:
This is still a WIP, filing early to get some feedback on direction. Tested locally with a multizone kind cluster with `TopologyAwareHints` feature gate enabled.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
- Enhancement Issue: https://github.com/kubernetes/enhancements/issues/2433

/sig network
/priority important-soon
/cc @aojea

### CODE REVIEW

File: test/e2e/network/topology_hints.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+			slices, listErr := c.DiscoveryV1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1.LabelServiceName, svc.Name)})
+			if listErr != nil {
+				return false, listErr
+			}
+
+			numEndpoints := 0
+			for _, slice := range slices.Items {
+				numEndpoints += len(slice.Endpoints)
+			}
+			if expectedPods > int32(numEndpoints) {
```
[the remark this answers is not in the document]
Comment (robscott) [reply test/e2e/network/topology_hints.go:139]: The flakes I'm remembering were caused by duplicate EndpointSlices getting created due to an out of date informer cache. Essentially the following sequence.

1. Pod A is created
2. EndpointSlice foo is created targeting Pod A
3. Pod B is created
4. Informer cache doesn't have EndpointSlice foo yet, so it creates EndpointSlice bar targeting Pods A and B

There's since been a number of fixes that should make that sequence much less likely, but I don't think it's completely impossible. The very nature of EndpointSlices means that consumers of the API need to dedup endpoints, so even if there are duplicate endpoints for some period of time, it should not break an implementation. All of that to say, this `>` instead of `!=` was actually intended to cover this potential edge case/flake. Hopefully that logic makes sense.

[the remark this answers is not in the document]
Comment (aojea) [reply test/e2e/network/topology_hints.go:139]: I'm fine if you are fine :smile:

File: test/e2e/network/topology_hints.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+var _ = common.SIGDescribe("Feature:Topology Hints", func() {
+	f := framework.NewDefaultFramework("topology-hints")
+
+	// filled in BeforeEach
+	var c clientset.Interface
+
+	ginkgo.BeforeEach(func() {
+		c = f.ClientSet
+		e2eskipper.SkipUnlessMultizone(c)
```
[the remark this answers is not in the document]
Comment (robscott) [reply test/e2e/network/topology_hints.go:48]: Asked in [#sig-testing Slack](https://kubernetes.slack.com/archives/C09QZ4DQB/p1635539368034000), no response yet. I've been running these locally with kind.

File: test/e2e/network/topology_hints.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+				return false, getErr
+			}
+			if ds.Status.NumberReady != ds.Status.DesiredNumberScheduled {
+				framework.Logf("Expected %d Pods to be ready, got %d", ds.Status.DesiredNumberScheduled, ds.Status.NumberReady)
+				return false, nil
+			}
+
+			expectedPods = ds.Status.NumberReady
+			return true, nil
+		})
```
[the remark this answers is not in the document]
Comment (robscott) [reply test/e2e/network/topology_hints.go]: Updated, thanks!

File: test/e2e/network/topology_hints.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+									SuccessThreshold: 1,
+									TimeoutSeconds:   1,
+								},
+							},
+						},
+					},
+				},
+			},
+			Status: appsv1.DaemonSetStatus{},
+		}
```
[the remark this answers is not in the document]
Comment (robscott) [reply test/e2e/network/topology_hints.go]: Updated, thanks!

**Comment by robscott** [comment]
In case anyone else is trying to run these tests locally, here's the kind config I'm using:

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
featureGates:
  TopologyAwareHints: true
nodes:
- role: control-plane
- role: worker
  kubeadmConfigPatches:
  - |
    kind: JoinConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        node-labels: "topology.kubernetes.io/zone=zone-a"
- role: worker
  kubeadmConfigPatches:
  - |
    kind: JoinConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        node-labels: "topology.kubernetes.io/zone=zone-b"
- role: worker
  kubeadmConfigPatches:
  - |
    kind: JoinConfiguration
    nodeRegistration:
      kubeletExtraArgs:
        node-labels: "topology.kubernetes.io/zone=zone-c"
```

File: test/e2e/network/topology_hints.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+			cmd := fmt.Sprintf("curl -q -s --connect-timeout 2 http://%s:80/", svc.Name)
+			consecutiveRequests := 0
+			var stdout string
+			if pollErr := wait.PollImmediate(framework.Poll, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
+				var err error
+				stdout, err = framework.RunHostCmd(f.Namespace.Name, execPod.Name, cmd)
+				if err != nil {
+					framework.Logf("unexpected error running %s from %s", cmd, execPod.Name)
+					return false, nil
+				}
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:168]: doing an "exec" per iteration is expensive and a source of flakiness,  it is more robust to build just one command that runs all the requests in a loop

https://github.com/kubernetes/kubernetes/blob/08bf54678e2bef8b4c72c8c6d104b0bdf19a6f75/test/e2e/network/service.go#L128-L138

I also will try to go for higher numbers of requests to be more resilient , 1 error from 5 is much problematic than 1 of 50 :smile:

Comment (robscott) [reply test/e2e/network/topology_hints.go]: Chatted about this in Slack and clarified that it's not so much a percentage of successful requests as it is a number of consecutive successful requests that this is looking for. The whole idea here is that it could take some time for kube-proxy to correctly program iptables rules. Once it does, we want to ensure that all requests are routed within the same zone.

@aojea had the great idea to read pod logs instead of either a complicated shell script or multiple kubectl execs. That's the approach I've updated the PR to use. In my limited testing, the full e2e run for this test is ~35% faster with this approach, and also likely much less prone to flakiness.

File: test/e2e/network/topology_hints.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (robscott) [reply test/e2e/network/topology_hints.go:49]: To clarify, I don't think getting these tests to run regularly in CI necessarily needs to block these tests from getting merged in. I think these already have appropriate guards (feature-prefixed name + e2eskipper.SkipUnlessMultizone) that should prevent this from running accidentally.

[the remark this answers is not in the document]
Comment (aojea) [reply test/e2e/network/topology_hints.go:49]: absolutely

File: test/e2e/network/topology_hints.go [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+			fromZone, ok := nodesByZone[nodeName]
+			if !ok {
+				framework.Failf("Expected zone to be specified for %s node", nodeName)
+			}
+			ginkgo.By("creating a client pod for probing the service from " + fromZone)
+			podName := "curl-from-" + fromZone
+			clientPod := e2epod.NewAgnhostPod(f.Namespace.Name, podName, nil, nil, nil, "serve-hostname")
+			nodeSelection := e2epod.NodeSelection{Name: nodeName}
+			e2epod.SetNodeSelection(&clientPod.Spec, nodeSelection)
+			cmd := fmt.Sprintf(`date; for i in $(seq 1 3000); do sleep 1; echo "Date: $(date) Try: ${i}"; curl -q -s --connect-timeout 2 http://%s:80/ ; echo; done`, svc.Name)
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:160]: does it server the hostname in the root url or in the `/hostname` ?

File: test/e2e/network/topology_hints.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+			clientPod.Spec.Containers[0].Command = []string{"/bin/sh", "-c", cmd}
+			clientPod.Spec.Containers[0].Name = clientPod.Name
+			f.PodClient().CreateSync(clientPod)
+
+			framework.Logf("Ensuring that requests from %s pod on %s node stay in %s zone", clientPod.Name, nodeName, fromZone)
+
+			var logs string
+			if pollErr := wait.PollImmediate(framework.Poll, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:168]: `framework.Poll = 2 * time.Second`  , since you do a curl per second with a 2 second timeout, maybe you should use 
```suggestion
			if pollErr := wait.Poll(5*time.Second, e2eservice.KubeProxyLagTimeout, func() (bool, error) {
```
to avoid polling at the beginning

File: test/e2e/network/topology_hints.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+				framework.ExpectNoError(err)
+				framework.Logf("Pod client logs: %s", logs)
+				logLines := strings.Split(logs, "\n")
+				consecutiveSameZone := 0
+
+				for i := len(logLines) - 1; i > 0; i-- {
+					if logLines[i] == "" || strings.HasPrefix(logLines[i], "Date:") {
+						continue
+					}
+					destZone, ok := podsByZone[logLines[i]]
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:180]: I don't know if you'll have to trim spaces, can't remember honestly,

File: test/e2e/network/topology_hints.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+					if !ok {
+						framework.Logf("could not determine dest zone from log line: %s", logLines[i])
+						return false, nil
+					}
+					if fromZone != destZone {
+						framework.Logf("expected request from %s to stay in %s zone, delivered to %s zone", clientPod.Name, fromZone, destZone)
+						return false, nil
+					}
+					consecutiveSameZone++
+					if consecutiveSameZone >= 5 {
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:190]: nit, `if  len(logLines) < 6` you can also return fast

File: test/e2e/network/topology_hints.go (hunk 7 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go]: Updated, thanks!

File: test/e2e/network/topology_hints.go (hunk 9 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:196]: Updated, thanks!

File: test/e2e/network/topology_hints.go (hunk 8 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:186]: Not currently necessary. Pod name is printed out without space or newline. I add a newline with the extra `echo` in the loop. Assuming this behavior won't change in the future.

File: test/e2e/network/topology_hints.go (hunk 6 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:160]: Serves pod name with current request.

**Comment by aojea** [comment]
/lgtm
/approve
/hold
I can not test it and the CI will not, so feel free to unhold if you are confident that is ok now, :upside_down_face:

**Comment by aojea** [comment]
once the test-infra job merges we should kick it with
/test pull-kubernetes-e2e-kind-multizone

**Comment by robscott** [comment]
From the output of that test run:

>{"msg":"PASSED [sig-network] Feature:Topology Hints should distribute endpoints evenly","total":417,"completed":40,"skipped":565,"failed":0}

Thanks for all the help with this! I think that's sufficient signal to merge this in.

/hold cancel

**Comment by aojea** [comment]
```
STEP: having hints set for each endpoint
STEP: keeping requests in the same zone
STEP: creating a client pod for probing the service from zone-a
Nov  4 22:32:56.422: INFO: The status of Pod curl-from-zone-a is Pending, waiting for it to be Running (with Ready = true)
Nov  4 22:32:58.425: INFO: The status of Pod curl-from-zone-a is Running (Ready = true)
Nov  4 22:32:58.427: INFO: Ensuring that requests from curl-from-zone-a pod on kind-worker node stay in zone-a zone
Nov  4 22:33:03.436: INFO: Pod client logs: Thu Nov  4 22:32:57 UTC 2021
Date: Thu Nov  4 22:32:58 UTC 2021 Try: 1
topology-serve-hostname-nsjz8
Date: Thu Nov  4 22:32:59 UTC 2021 Try: 2
topology-serve-hostname-nsjz8
Date: Thu Nov  4 22:33:00 UTC 2021 Try: 3
topology-serve-hostname-nsjz8
Date: Thu Nov  4 22:33:01 UTC 2021 Try: 4
topology-serve-hostname-nsjz8
Date: Thu Nov  4 22:33:02 UTC 2021 Try: 5
topology-serve-hostname-nsjz8
Date: Thu Nov  4 22:33:03 UTC 2021 Try: 6
topology-serve-hostname-nsjz8
STEP: creating a client pod for probing the service from zone-b
Nov  4 22:33:03.444: INFO: The status of Pod curl-from-zone-b is Pending, waiting for it to be Running (with Ready = true)
Nov  4 22:33:05.447: INFO: The status of Pod curl-from-zone-b is Running (Ready = true)
Nov  4 22:33:05.449: INFO: Ensuring that requests from curl-from-zone-b pod on kind-worker2 node stay in zone-b zone
Nov  4 22:33:10.460: INFO: Pod client logs: Thu Nov  4 22:33:04 UTC 2021
Date: Thu Nov  4 22:33:05 UTC 2021 Try: 1
topology-serve-hostname-hqbdx
Date: Thu Nov  4 22:33:06 UTC 2021 Try: 2
topology-serve-hostname-hqbdx
Date: Thu Nov  4 22:33:07 UTC 2021 Try: 3
topology-serve-hostname-hqbdx
Date: Thu Nov  4 22:33:08 UTC 2021 Try: 4
topology-serve-hostname-hqbdx
Date: Thu Nov  4 22:33:09 UTC 2021 Try: 5
topology-serve-hostname-hqbdx
Date: Thu Nov  4 22:33:10 UTC 2021 Try: 6
topology-serve-hostname-hqbdx
STEP: creating a client pod for probing the service from zone-c
Nov  4 22:33:10.467: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Nov  4 22:33:12.470: INFO: The status of Pod curl-from-zone-c is Running (Ready = true)
Nov  4 22:33:12.472: INFO: Ensuring that requests from curl-from-zone-c pod on kind-worker3 node stay in zone-c zone
Nov  4 22:33:17.493: INFO: Pod client logs: Thu Nov  4 22:33:11 UTC 2021
Date: Thu Nov  4 22:33:12 UTC 2021 Try: 1
topology-serve-hostname-xx87p
Date: Thu Nov  4 22:33:13 UTC 2021 Try: 2
topology-serve-hostname-xx87p
Date: Thu Nov  4 22:33:14 UTC 2021 Try: 3
topology-serve-hostname-xx87p
Date: Thu Nov  4 22:33:15 UTC 2021 Try: 4
topology-serve-hostname-xx87p
Date: Thu Nov  4 22:33:16 UTC 2021 Try: 5
topology-serve-hostname-xx87p
Date: Thu Nov  4 22:33:17 UTC 2021 Try: 6
```
nice log
